server/extension/extension.go

```diff
@@ -12,24 +11,147 @@ import (
+	//     "<apiVersion>:<kind>:<metadata.name>"
+	// Example:
+	//     Argocd-Resource-GVK-Name: "apps/v1:Pod:some-pod"
+	HeaderArgoCDResourceGVKName = "Argocd-Resource-GVK-Name"
 )
 
+// RequestResources defines the authorization scope for
+// an incoming request to a given extension. This struct
+// is populated from pre-defined Argo CD headers.
+type RequestResources struct {
+	ApplicationName      string
+	ApplicationNamespace string
+	ProjectName          string
+	Resources            []Resource
+}
+
+// Resource defines the Kubernetes resource used for checking
+// the authorization scope.
+type Resource struct {
+	Gvk  schema.GroupVersionKind
+	Name string
+}
+
+func (r Resource) String() string {
+	return fmt.Sprintf("%s, Name=%s", r.Gvk.String(), r.Name)
+}
+
+// ValidateHeaders will validate the pre-defined Argo CD
+// request headers for extensions and extract the resources
+// information populating and returning a RequestResources
+// object.
+// The pre-defined headers are:
+// - Argocd-Application-Name
+// - Argocd-Project-Name
+// - Argocd-Resource-GVK-Name
+//
+// The headers expected format is documented in each of the constant
+// types defined for them.
+func ValidateHeaders(r *http.Request) (*RequestResources, error) {
+	appHeader := r.Header.Get(HeaderArgoCDApplicationName)
+	if appHeader == "" {
+		return nil, fmt.Errorf("header %q must be provided", HeaderArgoCDApplicationName)
+	}
+	appNamespace, appName, err := getAppName(appHeader)
+	if err != nil {
+		return nil, fmt.Errorf("error getting app details: %s", err)
+	}
+	projName := r.Header.Get(HeaderArgoCDProjectName)
+	if projName == "" {
+		return nil, fmt.Errorf("header %q must be provided", HeaderArgoCDProjectName)
+	}
+	resources := []Resource{}
+	resourcesHeader := r.Header.Get(HeaderArgoCDResourceGVKName)
+	if resourcesHeader != "" {
+		resourceList, err := getResourceList(resourcesHeader)
+		if err != nil {
+			return nil, fmt.Errorf("error getting resource: %s", err)
+		}
+		resources = append(resources, resourceList...)
+	}
+
+	return &RequestResources{
+		ApplicationName:      appName,
+		ApplicationNamespace: appNamespace,
+		ProjectName:          projName,
+		Resources:            resources,
+	}, nil
+}
+
+func getAppName(appHeader string) (string, string, error) {
+	parts := strings.Split(appHeader, ":")
+	if len(parts) != 2 {
+		return "", "", fmt.Errorf("invalid value for %q header: expected format: <namespace>:<app-name>", HeaderArgoCDApplicationName)
+	}
+	return parts[0], parts[1], nil
+}
+
+func getResourceList(resourcesHeader string) ([]Resource, error) {
+	resources := []Resource{}
+	parts := strings.Split(resourcesHeader, ",")
+	for _, part := range parts {
+		resource, err := getResource(part)
+		if err != nil {
+			return nil, fmt.Errorf("resource error: %s", err)
+		}
+		if resource != nil {
+			resources = append(resources, *resource)
+		}
+	}
+
+	return resources, nil
+}
+
+func getResource(resourceString string) (*Resource, error) {
```

Comment:
Do we use this format elsewhere? If so, can we centralize the parsing logic?

Reply:
I don't believe this format is used anywhere in Argo CD. This is specific to the header as I didn't want to clash with the `/` char.